### ISSUE TITLE
Path to output files is hardcoded

### ISSUE THREAD

**Issue by dpaetzel** [stan-dev/cmdstanpy#628]
Hi! Thank you for creating and maintaining the `cmdstanpy` library! :slightly_smiling_face:

#### Summary:

The path to output files (`.hpp` as well as the compiled binary) seems to be hardcoded.

#### Description:

If I'm not mistaken, [this line](https://github.com/stan-dev/cmdstanpy/blob/develop/cmdstanpy/model.py#L490) hardcodes the path to the directory which will contain the generated `.hpp` file to be the same as the path to the directory of the `.stan` file. The same goes for the generated binary. This means that if I were to write a Python package which contains a Stan model as a `.stan` file and uses `cmdstanpy` to compile said model, then upon compiling that model, additional files (i.e. the binary and the `.hpp` file) are added to the installation directory of that package which is probably not what is intended.

This is especially problematic on systems that use rather strict declarative package management (e.g.  [NixOS](https://nixos.org)) where the installation directory of packages is read-only. In those cases, `cmdstanpy` fails to build the model because it cannot write the `.hpp` file to the place where it wants to.

I initially thought that uglily adding a combination of `stanc_options` and `os.getcwd()` could fix that, but it doesn't (presumably because the `--o` option of `stanc` is not being respected). I.e.

```
self.model_ = cmdstanpy.CmdStanModel(
    stan_file=model_filename,
    stanc_options={"--o" : os.path.join(os.getcwd(), "model.hpp")})
```

does not work.

#### Current Version:

```
INSTALLED VERSIONS
---------------------
python: 3.9.12 (main, Mar 23 2022, 21:36:19) 
[GCC 11.3.0]
python-bits: 64
OS: Linux
OS-release: 5.18.4
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
cmdstan_folder: /nix/store/kdscak71648wlf2qpld0jjqrbik3fw12-cmdstan-2.29.2/opt/cmdstan
cmdstan: (2, 29)
cmdstanpy: 1.0.7
pandas: 1.4.2
xarray: 2022.3.0
tqdm: 4.64.1
numpy: 1.21.5
ujson: 5.5.0
```

**Comment by WardBrian** [comment]
> added to the installation directory of that package which is probably not what is intended.

Why would this not be intended? Where else should these files live, and how would recommend that being requested by the user?

**Comment by dpaetzel** [comment]
Thank you for responding!

After installation of a package is finished, declarative package managers don't allow the installation directory to be touched any more. This means that right now I guess I'd have to add something along the lines of `cmdstanpy.Model(…).compile()` to the installation script of my package (with proper setup of dependencies first etc.); so I think I can somehow work around the problem. I was just surprised that `--o` doesn't get passed through to `stanc` via `stanc_options`.

My personal preference regarding this would be that my Python package provides the `.stan` file and upon users using the library and building a model, an `.hpp` file and a binary are created (e.g. in a hidden subdirectory of the user's current directory). However, I may prefer that because I'd like to keep my library's installation script simple.

Feel free to close this issue if you don't see the necessity to address this. Thank you for your time!

**Comment by ahartikainen** [comment]
Maybe you could use similar approach as `prophet` uses?

**Comment by WardBrian** [comment]
Yes, I think that the package install should handle this if it can. See prophet as @ahartikainen mentions, or https://github.com/WardBrian/cookiecutter-cmdstanpy-wrapper for ideas

On `--o`, this would not have the intended effect if we did pass it to stanc, since all it does is change the location of the `.hpp`, _not_ the final executable. This would be more confusing than anything, I think

**Comment by dpaetzel** [comment]
> On --o, this would not have the intended effect if we did pass it to stanc, since all it does is change the location of the .hpp, not the final executable. This would be more confusing than anything, I think

You're right, I didn't see that. Thank you for the hint regarding the prophet and cookiecutter projects!
